src/core/lib/support/function.h

```diff
@@ -0,0 +1,293 @@
+    return *this;
+  }
+
+  Function& operator=(Function&& rhs) {
+    if (this == &rhs) return *this;
+    vtable_->destruct(&storage_);
+    vtable_ = rhs.vtable_;
+    vtable_->move_construct(&rhs.storage_, &storage_);
+    return *this;
+  }
+
+  R operator()(Args... args) {
+    return vtable_->invoke(&storage_, std::forward<Args>(args)...);
+  }
+
+ private:
+  const function_impl::VTable<R, Args...>* vtable_;
+  typename std::aligned_storage<kInplaceStorage>::type storage_;
+};
+
+// InplaceFunction<> is like Function<>, but causes a compile time error if the
+// contained functor is bigger than kInplaceStorage
+// This class is guaranteed never to allocate
+template <typename R, typename... Args, size_t kInplaceStorage>
+class InplaceFunction<R(Args...), kInplaceStorage> {
+ public:
+  template <class F>
+  InplaceFunction(F&& f) {
+    static_assert(
+        sizeof(F) <= kInplaceStorage,
+        "InplaceFunction functor must be smaller than kInplaceStorage");
+    vtable_ = function_impl::Traits<R(Args...), F, false>::Construct(
+        std::forward<F>(f), &storage_);
+  }
+
+  InplaceFunction(const InplaceFunction& rhs) : vtable_(rhs.vtable_) {
+    vtable_->copy_construct(&rhs.storage_, &storage_);
+  }
+
+  InplaceFunction(InplaceFunction& rhs) : vtable_(rhs.vtable_) {
+    vtable_->copy_construct(&rhs.storage_, &storage_);
+  }
+
+  InplaceFunction(InplaceFunction&& rhs) : vtable_(rhs.vtable_) {
+    vtable_->move_construct(&rhs.storage_, &storage_);
+  }
+
+  ~InplaceFunction() { vtable_->destruct(&storage_); }
+
+  InplaceFunction& operator=(const InplaceFunction& rhs) {
+    if (this == &rhs) return *this;
+    vtable_->destruct(&storage_);
+    vtable_ = rhs.vtable_;
+    vtable_->copy_construct(&rhs.storage_, &storage_);
+    return *this;
+  }
+
+  InplaceFunction& operator=(InplaceFunction&& rhs) {
+    if (this == &rhs) return *this;
+    vtable_->destruct(&storage_);
+    vtable_ = rhs.vtable_;
+    vtable_->move_construct(&rhs.storage_, &storage_);
+    return *this;
+  }
+
+  R operator()(Args... args) {
+    return vtable_->invoke(&storage_, std::forward<Args>(args)...);
+  }
+
+ private:
+  const function_impl::VTable<R, Args...>* vtable_;
+  typename std::aligned_storage<kInplaceStorage>::type storage_;
+};
+
+namespace function_impl {
+// TODO(ctiller): use type_traits versions of these when compiler support is
+// available
+#if defined(__has_extension)
+#if __has_extension(has_trivial_copy) && __has_extension(has_trivial_destructor)
+template <typename T>
+struct is_trivially_copyable {
+  static const constexpr bool value = __has_trivial_copy(T);
+};
+template <typename T>
+struct is_trivially_destructable {
+  static const constexpr bool value = __has_trivial_destructor(T);
+};
+#define GRPC_INTERNAL_FUNCTION_USING_COMPILER
+#endif
+#endif
+#ifndef GRPC_INTERNAL_FUNCTION_USING_COMPILER
+//#warning Using subsitute type_trait functions
+// assume that other compilers will check this similarly enough
```

Comment:
Do we know which compilers will trigger this? Anything that we test against?

Reply:
gcc (up until surprisingly recently) lacks support, clang has it.

Can't actually turn on the warning (since builds will fail), but given we're compiling on at least one compiler that does the check I'm sort of ok here.